server/src/main/java/org/elasticsearch/rest/action/document/RestGetSourceAction.java

```diff
@@ -75,24 +76,46 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
                 validationError.addValidationError("fetching source can not be disabled");
                 channel.sendResponse(new BytesRestResponse(channel, validationError));
             } else {
-                client.get(getRequest, new RestResponseListener<GetResponse>(channel) {
-                    @Override
-                    public RestResponse buildResponse(final GetResponse response) throws Exception {
-                        final XContentBuilder builder = channel.newBuilder(request.getXContentType(), false);
-                        // check if doc source (or doc itself) is missing
-                        if (response.isSourceEmpty()) {
-                            return new BytesRestResponse(NOT_FOUND, builder);
-                        } else {
-                            final BytesReference source = response.getSourceInternal();
-                            try (InputStream stream = source.streamInput()) {
-                                builder.rawValue(stream, XContentHelper.xContentType(source));
-                            }
-                            return new BytesRestResponse(OK, builder);
-                        }
-                    }
-                });
+                client.get(getRequest, new RestGetSourceResponseListener(channel, request));
             }
         };
     }
 
+    static class RestGetSourceResponseListener extends RestResponseListener<GetResponse> {
+        private final RestRequest request;
+
+        RestGetSourceResponseListener(RestChannel channel, RestRequest request) {
+            super(channel);
+            this.request = request;
+        }
+
+        @Override
+        public RestResponse buildResponse(final GetResponse response) throws Exception {
+            checkIfSourceEmpty(response);
+
+            final XContentBuilder builder = channel.newBuilder(request.getXContentType(), false);
+            final BytesReference source = response.getSourceInternal();
+            try (InputStream stream = source.streamInput()) {
+                builder.rawValue(stream, XContentHelper.xContentType(source));
+            }
+            return new BytesRestResponse(OK, builder);
+        }
+
+        /**
+         * Checks if the requested source or document itself is missing.
+         *
+         * @param response a response
+         * @throws ResourceNotFoundException if source or doc itself is missing
+         */
+        private void checkIfSourceEmpty(final GetResponse response) {
+            if (response.isSourceEmpty()) {
```

Comment:
I'd prefer to check `isExists` first because, at least in my head, `isSourceEmpty` doesn't really make sense if the document doesn't exist. I know it *works* to do it this way, but it'd make me feel better to switch the order.

Reply:
You're totally right.